### PR TITLE
Simplify model configs

### PR DESCRIPTION
**Pull request by ojh31**
In anticipation for lots of new models when investigating scaling laws, I have taken the opportunity to simplify the model configs:

- Migrated config dicts to yaml
- Added memory requirements to model config
- Set model_parallel, gradient_checkpointing programatically based on size
- Moved default minibatch size to model config class
- Added docstring to model config class

### CODE REVIEW

**Comment by AlexTMallen** [comment]
I like this change! One thing I'll note is that memory requirements vary depending on the context length of the dataset, so we will need to do some nudging of that depending on dataset

File: weak_to_strong/config.py [hunk 1]
```diff
@@ -1,193 +1,116 @@
...
+        memory (float): 
+            The memory required for the model in bytes.
+        default_lr (float, optional):
+            The default learning rate. Defaults to 1e-5.
+        eval_batch_size (int, optional):
+            The batch size for evaluation. Defaults to 32.
+        minibatch_size_per_device (int, optional):
+            The minibatch size per device. Defaults to None.
+        lora_modules (list[str], optional):
+            The list of LORA modules. Defaults to None.
```
Comment (AlexTMallen) [inline weak_to_strong/config.py:43]: Thanks for adding the docstring! One other detail that should be added here is that LoRA will be used IFF `lora_modules` is not None

File: weak_to_strong/config.py [hunk 2]
```diff
@@ -1,193 +1,116 @@
...
+        custom_kwargs (dict, optional):
+            Custom keyword arguments. Defaults to None.
```
Comment (AlexTMallen) [inline weak_to_strong/config.py:45]: These kwargs are passed to the model.from_pretrained call

File: weak_to_strong/config.py [hunk 3]
```diff
@@ -1,193 +1,116 @@
...
+        gradient_checkpointing (bool, optional):
+            Whether to use gradient checkpointing. Defaults to None.
+        model_parallel (bool, optional):
+            Whether to use model parallelism. Defaults to None.
```
Comment (AlexTMallen) [inline weak_to_strong/config.py:49]: Model parallelism uses accelerate's automatic model sharding, while if model-parallel is false and you're using multiple GPUs, it uses data parallelism

File: weak_to_strong/config.py [hunk 4]
```diff
@@ -1,193 +1,116 @@
...
+        default_optimizer (str, optional):
+            The default optimizer. Defaults to "adam".
+    """
+
+    CHECKPOINTING_MEMORY = 3e9
+    MODEL_PARALLEL_FACTOR = 2
```
Comment (AlexTMallen) [inline weak_to_strong/config.py:55]: It would be good to also include a factor for max context length.

I'm wondering why this factor is so large--what does the `memory` parameter mean?

File: weak_to_strong/config.py [hunk 5]
```diff
@@ -1,193 +1,116 @@
...
+            )
+        if gradient_checkpointing is None:
+            gradient_checkpointing = memory > self.CHECKPOINTING_MEMORY
+        if minibatch_size_per_device is None:
+            minibatch_size_per_device = eval_batch_size
+        if not torch.cuda.is_bf16_supported() and (
+            custom_kwargs.get("torch_dtype") == "torch.bfloat16"
+        ):
+            custom_kwargs["torch_dtype"] = "torch.float32"
+        if not torch.cuda.is_bf16_supported() and custom_kwargs.get("bf16"):
```
Comment (AlexTMallen) [inline weak_to_strong/config.py:96]: I think we can get rid of this block because the first block is equivalent with appropriate yaml parameters (and modify the yamls accordingly). See [this PR](https://github.com/openai/weak-to-strong/commit/c01b150515581cb4f18bb59cadd677cecc207c14)

Comment (AlexTMallen) [reply weak_to_strong/config.py:96]: We should probably be using bf16 for more (all?) models since we're using LoRA

**Review by AlexTMallen — CHANGES_REQUESTED** [review]
Looks good overall! Just minor requests:

1. Switch from "bf16" and "fp32" to using the "torch_dtype" arg exclusively.
2. Tweak the docstring as mentioned
3. Account for the varying memory footprint based on context length size (and LoRA, though I wouldn't spend too much time on this since we don't really intend to use full finetuning). Accounting for context length might involve dynamically modifying a ModelConfig object before instantiating a model based on the passed max context length argument to train_simple, and the relationship between context length and memory should be proportional.

File: weak_to_strong/config.py (hunk 4 above)
Comment (ojh31) [reply weak_to_strong/config.py:55]: You think 2 is too large? I set the memory values here by looking at pytorch_model.bin sizes on huggingface. Ultimately though it's an arbitrary, configurable parameter which affects whether we use model parallelism, gradient checkpointing and bfloat16 by default.

File: weak_to_strong/config.py (hunk 5 above)
Comment (ojh31) [reply weak_to_strong/config.py]: > I think we can get rid of this block because the first block is equivalent with appropriate yaml parameters (and modify the yamls accordingly). See [this PR](https://github.com/openai/weak-to-strong/commit/c01b150515581cb4f18bb59cadd677cecc207c14)

In that case, this is much simpler, thank you!

**Review by AlexTMallen — APPROVED** [review]
